### PR TITLE
Use schema when parsing csv

### PR DESCRIPTION
**Pull request by vvijay-bolt**
## What changes are proposed in this pull request?

Use the input schema when parsing csv input during model serving. This ensures that the resulting dataframe has the correct types.

## How is this patch tested?

Train and serve the modified logistic regression example below.
Save the following in test_csv.csv:
x
1
-1

Then try:
curl -d '{"columns":["x"],"index":[0,1],"data":[[1],[-1]]}' -H 'Content-Type: application/json' localhost:5000/invocations
curl --data-binary '@test_csv.csv' -H 'Content-Type: text/csv' localhost:5000/invocations

The json request returns [1, 0].
On master, the csv request will produce: {"error_code": "BAD_REQUEST", "message": "Incompatible input types for column x. Can not safely convert int64 to int32."}
After applying the fix, the csv request will also product [1, 0].

Modified logistic regression example:

import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from mlflow.models.signature import infer_signature

import mlflow
import mlflow.sklearn

if __name__ == "__main__":
    X = pd.DataFrame({'x': [-2, -1, 0, 1, 2, 1]}, dtype='int32')
    y = np.array([0, 0, 1, 1, 1, 0])
    lr = LogisticRegression()
    lr.fit(X, y)
    score = lr.score(X, y)
    print("Score: %s" % score)

    class Model(mlflow.pyfunc.PythonModel):
        def __init__(self, m):
            self.m = m

        def predict(self, context, x):
            return self.m.predict(x)

    model = Model(lr)
    signature = infer_signature(X, model.predict(None, X))
    print('Signature')
    print(signature)
    mlflow.pyfunc.log_model('model', python_model=model, signature=signature)
    print("Model saved in run %s" % mlflow.active_run().info.run_uuid)

## Does this PR change the documentation?

No

### Is this a user-facing change?

No
### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [- ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes

### CODE REVIEW

**Comment by BenWilson2** [comment]
Hi @vvijay-bolt could you add a test suite to validate this change to ensure that the parsing of a .csv with different submitted data type orientations functions correctly? 
It can be added here: https://github.com/mlflow/mlflow/blob/master/tests/pyfunc/test_scoring_server.py

**Comment by vvijay-bolt** [comment]
@BenWilson2 I added a test (modeled after a similar json test) to verify that the datatypes were set correctly after parsing. The pandas functions to read and write csv files were throwing errors on the datetime fields, so I left them out of the test.

File: mlflow/utils/proto_json_utils.py [hunk 1]
```diff
@@ -193,6 +193,20 @@ def default(self, o):  # pylint: disable=E0202
             return super().default(o)
 
 
+def _extract_pandas_input_types_from_schema(schema):
```
Comment (arjundc-db) [inline mlflow/utils/proto_json_utils.py:196]: If the input is of type csv, it has to be a pandas input. Since a pandas schema will always be a ColSpec, I dont think this entire function is necessary.
Simply using `dtypes = dict(zip(schema.input_names(), schema.pandas_types()))` should suffice.

Comment (vvijay-bolt) [reply mlflow/utils/proto_json_utils.py]: Made the change.

File: mlflow/utils/proto_json_utils.py [hunk 2]
```diff
@@ -210,17 +210,7 @@ def _dataframe_from_json(
...
-            #  - a single tensor: attempt to parse all columns with the same dtype
-            #  - a dictionary of tensors: each column gets the type from an equally named tensor
-            if len(schema.inputs) == 1:
-                dtypes = schema.numpy_types()[0]
-            else:
-                dtypes = dict(zip(schema.input_names(), schema.numpy_types()))
-        else:
-            dtypes = dict(zip(schema.input_names(), schema.pandas_types()))
-
+        dtypes = _extract_pandas_input_types_from_schema(schema)
```
Comment (arjundc-db) [inline mlflow/utils/proto_json_utils.py:213]: Can you also revert this change please.

Comment (vvijay-bolt) [reply mlflow/utils/proto_json_utils.py]: Done.

**Comment by vvijay-bolt** [comment]
autoformat

**Review by arjundc-db — APPROVED** [review]
LGTM.
Thank you @vvijay-bolt !

**Review by tomasatdatabricks — APPROVED** [review]
lgtm!

File: mlflow/pyfunc/scoring_server/__init__.py [hunk 3]
```diff
@@ -136,14 +136,20 @@ def parse_json_input(json_input, orient="split", schema: Schema = None):
...
+def parse_csv_input(csv_input, schema: Schema = None):
     """
     :param csv_input: A CSV-formatted string representation of a Pandas DataFrame, or a stream
                       containing such a string representation.
+    :param schema: Optional schema specification to be used during parsing.
     """
 
     try:
-        return pd.read_csv(csv_input)
+        csv_input = StringIO(csv_input)
```
Comment (arjundc-db) [inline mlflow/pyfunc/scoring_server/__init__.py:147]: Let's move this line back to  https://github.com/mlflow/mlflow/pull/5531/files#diff-e607d3e6aefa036b1bfe9416fd1aa8b0934120f458cf263ddebab224f832e4a5L264 since the input to parse_csv can also be a path to a csv file. We need to fix our documentation for this.

File: tests/pyfunc/test_scoring_server.py [hunk 4]
```diff
@@ -429,6 +444,14 @@ def test_split_oriented_json_to_df():
     assert set(str(dt) for dt in df.dtypes) == {"object", "float64", "int64"}
 
 
+def test_parse_with_schema_csv(pandas_df_with_csv_types):
+    schema = Schema([ColSpec(c, c) for c in pandas_df_with_csv_types.columns])
+    df = _shuffle_pdf(pandas_df_with_csv_types)
+    csv_str = df.to_csv(index=False)
+    df = pyfunc_scoring_server.parse_csv_input(csv_str, schema=schema)
```
Comment (arjundc-db) [inline tests/pyfunc/test_scoring_server.py:451]: We need StringIO here:
```suggestion
    df = pyfunc_scoring_server.parse_csv_input(csv_str, schema=schema)
```

Please import it using 
```
try:
    from StringIO import StringIO
except ImportError:
    from io import StringIO
```
to maintain compatibility.

**Review by arjundc-db — COMMENTED** [review]
Sorry, two more comments to fix test failures and we should we good to go.

File: test_csv.csv [hunk 5]
```diff
@@ -0,0 +1,3 @@
+x
```
Comment (BenWilson2) [inline test_csv.csv:1]: Where is this used?

File: examples/sklearn_logistic_mod/train.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import numpy as np
```
Comment (BenWilson2) [inline examples/sklearn_logistic_mod/train.py]: Any new examples need to be registered to execute as part of the Examples CI here: https://github.com/mlflow/mlflow/blob/101ad6e8eb383c769178df0df83d1d2a1cea6b4a/tests/examples/test_examples.py#L92-L94
. Do we need this additional test or can a signature example be added to examples/sklearn_logistic_regression/train.py? 
It might also be worthwhile to add a load from the logged (saved) location to run the score inference again to demonstrate the loading of the pyfunc flavor with the type-validated signature for a .csv file.

**Review by BenWilson2 — APPROVED** [review]
LGTM!
